csrc/lower_thread_predicate.cpp

```diff
@@ -494,6 +519,224 @@ class RedundantUseAnalysis : BackwardVisitor {
+  // step-1, check if there is a leaf domain parallelized by thread/block id and
+  // its definition is a merged
+  bool merged_parallelized_thread_block = false;
+  for (auto ld : out_tv->domain()->domain()) {
+    const ParallelType& pt = ld->getParallelType();
+    auto merge = dynamic_cast<Merge*>(ld->definition());
+    if (isParallelTypeThread(pt) && merge) {
+      merged_parallelized_thread_block = true;
+      break;
+    }
+  }
+  // shortcut if there is no such leaf domain
+  if (!merged_parallelized_thread_block) {
+    return;
+  }
+
+  const auto& root_domain = out_tv->getMaybeRFactorDomain();
+  // For each broadcast root domain, find a concretized domain from its
+  // exact mapped domain set.
+  auto getConcretizedBroadcastRootDomain = [&]() {
+    std::unordered_map<IterDomain*, IterDomain*>
+        concretized_broadcast_root_domains;
+    for (auto rd : root_domain) {
+      if (rd->isBroadcast()) {
+        const auto& mapped_rd_exact_set = GpuLower::current()
+                                              ->caMap()
+                                              ->getIdSets(IdMappingMode::EXACT)
+                                              .getDisjointSetOf(rd);
+        for (auto mapped_rd : mapped_rd_exact_set) {
+          std::unordered_set<IterDomain*> all_cids =
+              GpuLower::current()
+                  ->concretizedBroadcastDomains()
+                  ->allConcretizedDomains(mapped_rd);
+          if (!all_cids.empty()) {
+            concretized_broadcast_root_domains[rd] = *all_cids.begin();
+            break;
+          }
+        }
+      }
+    }
+    return concretized_broadcast_root_domains;
+  };
+
+  // Find all the root domains that are merged to the leaf domain.
+  // e.g. Root: [I1,B2,B3] -> Leaf: [I1B2B3], the merged_root_domains =
+  // {B3,B2,I1}
+  auto getRootDomainsMergedToLeaf = [&](IterDomain* ld) {
+    std::vector<IterDomain*> merged_root_domains;
+    std::vector<int> index_root_domain;
+    std::vector<IterDomain*> intermediate_domains = root_domain;
+    auto all_exp = DependencyCheck::getAllExprsBetween(
+        {root_domain.begin(), root_domain.end()}, {ld});
+    for (auto expr : all_exp) {
+      if (auto merge = dynamic_cast<Merge*>(expr)) {
+        auto outer_iter_im = std::find(
+            intermediate_domains.begin(),
+            intermediate_domains.end(),
+            merge->outer());
+        auto inner_iter_im = std::find(
+            intermediate_domains.begin(),
+            intermediate_domains.end(),
+            merge->inner());
+        TORCH_INTERNAL_ASSERT(
+            outer_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->outer());
+        TORCH_INTERNAL_ASSERT(
+            inner_iter_im != intermediate_domains.end(),
+            "Couldn't find ",
+            merge->inner());
+        int dist = std::distance(inner_iter_im, outer_iter_im);
+        if (std::abs(dist) != 1) {
```

Comment:
I'm skeptical that changing the position of merges would work as it would change the indexing.
`T1[I0, I1]->merge(0, 1)`
is not indexed the same as
`T1[I0, I1]->merge(1, 0)`

Reply:
Does line 633 indicate that the predicate is invariant to the order of the root domains going into the merge?

Reply:
`order of the root domains going into the merge` doesn't influence the indexing. Here, we are indexing between multi-dimension neighboring root domains (indices) and one leaf domain (linear_index). e.g. `[I0, I1] --> [I0*I1], linear_index = indices[0]*stride[0] + indices[1]*stride[1]; stride=[len(I1), 1]; The range of indices[i] is from 0 to len(i)-1`